AWSManagedMachinePool - Launch Template support

richardchen331 opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add launch template support to `AWSManagedMachinePool`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2055

**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. 
2. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE"....however we encourage contributors to never use this as release notes are incredible useful.
-->
```release-note
Add launch template support to `AWSManagedMachinePool`.
```

richardcase commented:
/ok-to-test

richardchen331 commented:
/retest

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

sedefsavas commented:
@richardchen331 @richardcase is this implementation in line with https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3024?

richardcase commented:
> @richardchen331 @richardcase is this implementation in line with #3024?

@sedefsavas - i need to update the proposal based on a chat that i had with @richardchen331. We decided on a way forward to unblock development. If after the proposal has been updated and changes are needed we can go back and refactor. I will check the proposal updated by EOD Wednesday.

richardcase commented on exp/api/v1beta1/awsmanagedmachinepool_types.go:
```diff
@@ -144,6 +143,9 @@ type AWSManagedMachinePoolSpec struct {
 	CapacityType *ManagedMachinePoolCapacityType `json:"capacityType,omitempty"`
+
+	// +optional
+	AWSLaunchTemplate *AWSLaunchTemplate `json:"awsLaunchTemplate,omitempty"`
```
It would be good to have a comment here. Maybe worth mentioning that you can't use this if you fields X, Y, Z?

richardchen331 replied:
Sounds good. Added comment here.

richardcase commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -182,6 +201,21 @@ func (r *AWSManagedMachinePoolReconciler) reconcileNormal(
+		}
+
+		// set the LaunchTemplateReady condition
+		conditions.MarkTrue(machinePoolScope.ManagedMachinePool, expinfrav1.LaunchTemplateReadyCondition)
```
We should probably mark this condition false in the above error scenarios, along with a reason.

richardchen331 replied:
That makes sense. Added logic to mark the condition to false in this case.

richardcase commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -192,15 +226,40 @@ func (r *AWSManagedMachinePoolReconciler) reconcileNormal(
+		}
+
+		machinePoolScope.Info("deleting launch template", "name", launchTemplate.Name)
+		if err := ec2Svc.DeleteLaunchTemplate(*launchTemplateID); err != nil {
```
Should the delete use the `launchTemplate.LaunchTemplateId` instead?

richardchen331 replied:
There's no LaunchTemplateId field in `AWSLaunchTemplate` at this moment, that's why i'm using launchTemplateID instead. Please let me know if that makes sense. Thanks!

richardcase commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -266,3 +325,187 @@ func managedControlPlaneToManagedMachinePoolMapFunc(c client.Client, gvk schema.
+	return ec2.NewService(scope)
+}
+
+func (r *AWSManagedMachinePoolReconciler) reconcileLaunchTemplate(machinePoolScope *scope.ManagedMachinePoolScope, ec2Scope scope.EC2Scope) error {
```
Is there anyway we could have a common function for this across [AWSMachinePool](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/833ca27c87eab8a8f7fd98ca85b46dcd1a22e94f/exp/controllers/awsmachinepool_controller.go#L383) and AWSManagedMachinePool?

richardcase replied:
Do we need to extend the ec2 service? Or perhaps refactor and create a launch template service?

richardchen331 replied:
Sounds good. I moved the logic into the launchtemplate service (`ReconcileLaunchTemplate`).

richardcase commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -266,3 +325,187 @@ func managedControlPlaneToManagedMachinePoolMapFunc(c client.Client, gvk schema.
+	return nil
+}
+
+func (r *AWSManagedMachinePoolReconciler) reconcileTags(machinePoolScope *scope.ManagedMachinePoolScope, ec2Scope scope.EC2Scope) error {
```
Same comment about reusing the logic between the 2 different machine pool controllers

richardchen331 replied:
Moved the logic into the launchtemplate service as well.

richardcase commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -266,3 +325,187 @@ func managedControlPlaneToManagedMachinePoolMapFunc(c client.Client, gvk schema.
+	return nil
+}
+
+func (r *AWSManagedMachinePoolReconciler) machinePoolAnnotationJSON(machinePool *expinfrav1.AWSManagedMachinePool, annotation string) (map[string]interface{}, error) {
```
This and the `machinePoolAnnotation` below feels like it could be useful on the scope. wdyt?

richardchen331 replied:
Sounds good. Moved the logic into launchtemplate scope.

sedefsavas commented:
Given that this is an important feature and changes the way we interact with `AWSManagedMachinePool`, we would want to have this in e2e tests and there is already a proposal out but would be good to update existing docs.

richardcase commented:
> Given that this is an important feature and changes the way we interact with `AWSManagedMachinePool`, we would want to have this in e2e tests and there is already a proposal out but would be good to update existing docs.

Agree on adding this to the e2e tests for eks. This is covered in the updated proposal. @richardchen331 i can guide you through this if you want?

richardcase commented on exp/api/v1beta1/awsmanagedmachinepool_types.go:
```diff
@@ -105,7 +105,6 @@ type AWSManagedMachinePoolSpec struct {
 
 	// AMIType defines the AMI type
 	// +kubebuilder:validation:Enum:=AL2_x86_64;AL2_x86_64_GPU;AL2_ARM_64
-	// +kubebuilder:default:=AL2_x86_64
```
We still need this default.

richardchen331 replied:
Got it. Added back the default.

richardchen331 commented:
> > Given that this is an important feature and changes the way we interact with `AWSManagedMachinePool`, we would want to have this in e2e tests and there is already a proposal out but would be good to update existing docs.
> 
> Agree on adding this to the e2e tests for eks. This is covered in the updated proposal. @richardchen331 i can guide you through this if you want?

@richardcase  Agree that we should have e2e tests for it. It would be great if you could guide me on how to add e2e tests for eks :)

richardcase commented:
> Agree that we should have e2e tests for it. It would be great if you could guide me on how to add e2e tests for eks :)

I'll ping you in slack and we can arrange a zoom sessions

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

pydctw commented:
As you created a new test flavor, it should be added to files section in https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/data/e2e_eks_conf.yaml#L179-L207.

richardchen331 commented:
> I was just reviewing this PR 😀 As you have created a new test flavor, it should be added to files section in https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/data/e2e_eks_conf.yaml#L179-L207.

@pydctw Thank you! Just updated the conf and i'll retry the test again :)

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
Hi @richardcase @sedefsavas , I added e2e test for this change. Could you help review it again? Thanks!

richardchen331 commented on test/e2e/data/eks/cluster-template-eks-managed-machinepool-with-launch-template-only.yaml:
```diff
@@ -0,0 +1,37 @@
+      version: "EKS_KUBERNETES_VERSION"
+      clusterName: "${CLUSTER_NAME}"
+      bootstrap:
+        dataSecretName: "${CLUSTER_NAME}-pool-0-userdata"
```
This would not work here. The existing behavior requires the user to specify a data secret used as bootstrap data in AWSMachinePool (https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/scope/machinepool.go#L126)

richardchen331 commented on test/e2e/data/eks/cluster-template-eks-managed-machinepool-with-launch-template-only.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachinePool
+metadata:
+  name: "${CLUSTER_NAME}-pool-0"
```
Got it. I updated the machine pool to a different name, does that sound good?

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

sedefsavas commented:
Is this ready for another round of reviews?

richardchen331 commented:
> Is this ready for another round of reviews?

@sedefsavas @richardcase  Yes it is. Could you help take another look? Thanks!

richardcase commented:
A couple of things to note:

- Added a follow-up issue to add support for the EKS bootstrapper for `AWSManagedMachinePools`. See #3363
- Added a note to handle the deprecated fields in v1beta2. See #2355

richardcase commented on exp/api/v1beta1/awsmanagedmachinepool_webhook.go:
```diff
@@ -188,6 +214,16 @@ func (r *AWSManagedMachinePool) validateImmutable(old *AWSManagedMachinePool) fi
 	appendErrorIfMutated(old.Spec.AMIType, r.Spec.AMIType, "amiType")
 	appendErrorIfMutated(old.Spec.RemoteAccess, r.Spec.RemoteAccess, "remoteAccess")
 	appendErrorIfSetAndMutated(old.Spec.CapacityType, r.Spec.CapacityType, "capacityType")
+	if (old.Spec.AWSLaunchTemplate != nil && r.Spec.AWSLaunchTemplate == nil) ||
```
Could we use the `appendErrorIfMutated` for this test?

richardchen331 replied:
I was thinking that we should allow user to modify the launch template (e.g. update AMI ID), but should not allow user to change a node group from using launch template to not using launch template (and vice versa). That's why I didn't use `appendErrorIfMutated`. Let me know if that makes sense. Thanks!

richardcase commented on exp/controllers/awsmachinepool_controller.go:
```diff
@@ -140,6 +139,21 @@ func (r *AWSMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	}
 
+	// Create the launch template scope
+	launchTemplateScope, err := scope.NewLaunchTemplateScope(scope.LaunchTemplateScopeParams{
```
Do we need a new scope at this level? We have the `MachinePoolScope` is the scope of the reconciliation of the machine pool, so personally i don't think we need it here. But there is nothing stopping us creating the scope further down.

richardcase replied:
I would potentially change `ec2Svc.ReconcileLaunchTemplate` so that it accepts the machine pool scope like we do for the [NodegroupService](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/eks/service.go#L72).

richardchen331 replied:
Good suggestion! I refactored the code so we don't create a new scope here and in the AWSMMP controller.

richardcase commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -150,15 +164,32 @@ func (r *AWSManagedMachinePoolReconciler) Reconcile(ctx context.Context, req ctr
 	}
 
+	// Create the launch template scope
+	launchTemplateScope, err := scope.NewLaunchTemplateScope(scope.LaunchTemplateScopeParams{
```
Same comment as previous around the introduction of a new scope as this level.

richardcase commented on pkg/cloud/scope/launchtemplate.go:
```diff
@@ -0,0 +1,202 @@
+
+const (
+	// AWSManagedControlPlaneKind is the Kind of AWSManagedControlPlane.
+	AWSManagedControlPlaneKind = "AWSManagedControlPlane"
```
We have this defined in the [machine controller](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0babdf15d5584744c4693e8be4ab44c580847531/controllers/awsmachine_controller.go#L79)...do you think we should move it somewhere shared?

richardchen331 replied:
I don't know what is a good place to put the constant.. Do you have any suggestions?

richardcase replied:
Thats a hard one. Perhaps we add it to the current EKS control plane API version, so `controlplane/eks/api/v1beta1`

richardcase commented on pkg/cloud/scope/launchtemplate.go:
```diff
@@ -0,0 +1,202 @@
+	return tags
+}
+
+func (m *LaunchTemplateScope) GetRawBootstrapData() ([]byte, error) {
```
We have this already on the [MachinePoolScope](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0babdf15d5584744c4693e8be4ab44c580847531/pkg/cloud/scope/machinepool.go#L126) so it feels like we should using that implementation and passing the result down into the service.

richardcase replied:
Same with the tags above

richardcase replied:
As a general pattern if we have the parent machine pool scope that kicks of reconciliation of another resource we can pass down the scope or the values instead of reimplemnting the functions.

richardchen331 replied:
That makes sense. Updated the code to reuse implementation from MachinePoolScope

richardcase commented:
@richardchen331 - just checking on this. How are you egtting on, did you need anything?

richardchen331 commented:
Hi @richardcase , thanks for checking in on this. I'm busy w/ work recently and I'll update the PR based on the feedback this week.

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
@richardcase I refactored the code based on your feedback. Could you help take another look? Thanks!

richardcase commented:
@richardchen331 - sorry for the delay in getting back to this. I will be doing another pass this week.

richardcase commented on pkg/cloud/scope/launchtemplate.go:
```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
```suggestion
Copyright 2022 The Kubernetes Authors.
```

richardchen331 replied:
Updated

richardcase commented on templates/cluster-template-eks-managedmachinepool-launchtemplate.yaml:
```diff
@@ -0,0 +1,161 @@
+    mapRoles:
+      - groups:
+          - system:masters
+        rolearn: arn:aws:iam::707343435239:role/aws-dev_databricks-view-only
```
Guessing these need to be removed from the template :smiley:

richardcase replied:
If you could add the envsubst declarations like the other templates that would be great.

richardchen331 replied:
oops, thanks for catching this! this is not intended and removed now

richardcase commented on test/e2e/suites/managed/machine_pool.go:
```diff
@@ -63,11 +70,25 @@ func ManagedMachinePoolSpec(ctx context.Context, inputGetter func() ManagedMachi
-	err := shared.ApplyTemplate(ctx, configCluster, input.BootstrapClusterProxy)
+	workloadClusterTemplate := shared.GetTemplate(ctx, configCluster)
+	if input.UsesLaunchTemplate {
+		userDataTemplate := `#!/bin/bash
```
Why do we need this? Does this not come from the bootstrap secret?

(i will go back and check the code ^^^^)

richardcase replied:
Created #3363  for this

richardcase reviewed (commented):
@richardchen331 - a few comments. Would you also be able to rebase and squash your commits so they tell a story?

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
@richardcase This PR is rebased and ready for review. Could you help take a look when you have time? Thanks!

richardcase commented:
Awesome thanks @richardchen331 :tada:  I will review first thing tomorrow morning. And great to see the e2e passing.

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
@richardcase could you kindly help take a look at this PR when you have time? it would be difficult to keep rebasing as the master branch changes... Thanks!

pydctw commented:
@richardchen331, I reviewed the PR a long time ago but can review again. 

Is the PR almost final? Asking because there is a plan to bump API to v1beta2 in near future and I think we should get this one merged prior if possible as this PR make many changes to v1beta1 API.

cc @sedefsavas @richardcase (out this week)

richardchen331 commented:
> @richardchen331, I reviewed the PR a long time ago but can review again.
> 
> Is the PR almost final? Asking because there is a plan to bump API to v1beta2 in near future and I think we should get this one merged prior if possible as this PR make many changes to v1beta1 API.
> 
> cc @sedefsavas @richardcase (out this week)

Thank you @pydctw ! Yes I have addressed all prior review comments already and it's ready for final review. Please feel free to reach out to me if you have any questions about this PR. Thanks again for reviewing!

pydctw commented on test/e2e/suites/managed/eks_test.go:
```diff
@@ -36,12 +36,13 @@ import (
+		clusterName          string
+		cniAddonName         = "vpc-cni"
+		corednsAddonName     = "coredns"
+		eksKubernetesVersion = "v1.21.14"
```
Can we use [KUBERNETES_VERSION](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/data/e2e_eks_conf.yaml#L177) from e2e_eks_conf.yaml instead of specifying one here?

MachinePool uses k8s version of v1.21.14 while AWSManagedControlPlane uses v1.22. See [artifacts](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/3094/pull-cluster-api-provider-aws-e2e-eks/1557132529587195904/artifacts/clusters/bootstrap/resources/cluster-fjcwdc/). Unless there is a specific reason for it, I think it will be great to be consistent.

richardchen331 replied:
This was added a long time ago. I think at that time using `KUBERNETES_VERSION` didn't work, but maybe things changed and let me try again.

pydctw replied:
Sounds great. This is not a blocker but a suggestion. If it takes too much time, we can refactor as a follow up.

richardchen331 replied:
I removed it and now uses `KUBERNETES_VERSION` instead, it's tested and works well.

pydctw commented on pkg/cloud/services/ec2/launchtemplate.go:
```diff
@@ -27,15 +28,296 @@ import (
+		return scope.PatchObject()
+	}
+
+	if needsUpdate || tagsChanged || *imageID != *launchTemplate.AMI.ID {
```
I am confused by the logic and trying to confirm my understanding. When does this `if` statement run?

`if needsUpdate || tagsChanged || *imageID != *launchTemplate.AMI.ID`  is true, doesn't it return on line 156 and never reaches this line?

richardchen331 replied:
Thanks for pointing it out! You're right and let me update the code here.

pydctw commented:
Is the plan to mark `AMIType,AMIVersion,InstanceType,DiskSize,InstanceProfile` fields as deprecated in this PR and remove in v1beta2? 

That is my understanding from the [proposal](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3365/files#diff-d6d8565c707995a7204010e1acade42238594c2bcf88d13a6ed7bd316369942eR244) but want to confirm.

richardcase commented:
Yes they should be marked as deprecated. Then when we bump the API version in the future we can remove them.....assuming we have suitable conversion functions in place.

richardchen331 commented on exp/controllers/awsmanagedmachinepool_controller.go:
```diff
@@ -198,15 +243,40 @@ func (r *AWSManagedMachinePoolReconciler) reconcileNormal(
+		}
+
+		if launchTemplate == nil {
+			machinePoolScope.V(2).Info("Unable to locate launch template")
```
done

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardchen331 commented on exp/api/v1beta1/awsmanagedmachinepool_webhook.go:
```diff
@@ -117,6 +117,26 @@ func (r *AWSManagedMachinePool) validateRemoteAccess() field.ErrorList {
+	if r.Spec.AWSLaunchTemplate.IamInstanceProfile != "" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "AWSLaunchTemplate", "IamInstanceProfile"), r.Spec.AWSLaunchTemplate.IamInstanceProfile, "IAM instance profile in launch template is prohibited in EKS managed node group"))
+	}
+
```
`AMIType` is still required when making calls to EKS using the [AWS SDK](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#CreateNodegroupInput). 
We could not allow people to specify `AMIType` and use hardcoded value of `CUSTOM` when `AWSLaunchTemplate` is specified, but I'm wondering if there would be use cases where user want to create a node group without using launch template.

richardcase replied:
@pydctw - i added this to the agenda for Monday's office hours to discuss.

richardchen331 commented:
@pydctw Thanks for all the feedbacks! I've updated the PR based on your comments. Could you take another look? Thanks!

pydctw commented:
Thanks for addressing all the comments, @richardchen331. It looks good to me. 

Two remaining items that are related regarding the usage of launchtemplate for AWSManagedMachinePool. Let's wait for @richardcase's opinion on this.
* [comment1](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#issuecomment-1211373213)
* [comment2](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#discussion_r946006474)

FYI, we are trying to include the PR in the next release.

richardchen331 commented:
> Thanks for addressing all the comments, @richardchen331. It looks good to me.
> 
> Two remaining items that are related regarding the usage of launchtemplate for AWSManagedMachinePool. Let's wait for @richardcase's opinion on this.
> 
> * [comment1](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#issuecomment-1211373213)
> * [comment2](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#discussion_r946006474)
> 
> FYI, we are trying to include the PR in the next release.

@richardcase @pydctw just want to follow up and see if we have made any decision about these two comments? Thanks!

pydctw commented:
> @richardcase @pydctw just want to follow up and see if we have made any decision about these two comments? Thanks!

Thanks for the reminder., @richardchen331. I planned to update the PR after the office hrs but forgot.

As there was no strong opinion whether to 1) Deprecate the existing fields and only support LaunchTemplate or 2) Support both existing and new way (LaunchTemplate), I propose we proceed as following.
* Merge the PR as is - it looks good and all test passed. 
* Discuss two options in the [proposal](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3365/files#diff-d6d8565c707995a7204010e1acade42238594c2bcf88d13a6ed7bd316369942eR117).
* Depending on the decision, create a new PR to mark those fields as deprecated if we decided on option 1, otherwise do nothing.

pydctw commented:
/lgtm

cc @richardcase for an approval

sedefsavas commented:
/lgtm

richardcase commented:
/approve

Ankitasw commented:
Can be moved to next API change.
cc @richardcase
